src/tribol/mesh/MfemData.cpp

```diff
@@ -891,6 +915,221 @@ std::unique_ptr<mfem::BlockOperator> MfemJacobianData::GetMfemBlockJacobian( con
+                                submesh_data_.GetSubmeshFESpace().Dof_TrueDof_Matrix() ) );
+
+  // transfer (1, 0) block (gap dof rows, displacement dof cols)
+  submesh_J = GetUpdateData().submesh_redecomp_xfer_10_->TransferToParallelSparse(
+      lm_elems, mortar_elems,
+      method_data.getBlockJ()( static_cast<int>( BlockSpace::LAGRANGE_MULTIPLIER ),
+                               static_cast<int>( BlockSpace::MORTAR ) ) );
+  submesh_J += GetUpdateData().submesh_redecomp_xfer_10_->TransferToParallelSparse(
+      lm_elems, nonmortar_elems,
+      method_data.getBlockJ()( static_cast<int>( BlockSpace::LAGRANGE_MULTIPLIER ),
+                               static_cast<int>( BlockSpace::NONMORTAR ) ) );
+  submesh_J.Finalize();
+  submesh_J_hypre = GetUpdateData().submesh_redecomp_xfer_10_->ConvertToHypreParMatrix( submesh_J, false );
+  // Matrix returned by mfem::ParMult copies row and column starts since last arg is true. All other data is copied and
+  // owned by the new matrix.
+  parent_J_hypre = std::unique_ptr<mfem::HypreParMatrix>( std::unique_ptr<mfem::HypreParMatrix>(
+      mfem::ParMult( submesh_J_hypre.get(), submesh_parent_vdof_xfer_.get(), true ) ) );
+  block_J->SetBlock( 1, 0,
+                     mfem::RAP( submesh_data_.GetSubmeshFESpace().Dof_TrueDof_Matrix(), parent_J_hypre.get(),
+                                parent_data_.GetParentCoords().ParFESpace()->Dof_TrueDof_Matrix() ) );
+
+  // Create ones on diagonal of eliminated mortar tdofs (CSR sparse matrix -> HypreParMatrix) I vector
+  auto& submesh_fes = submesh_data_.GetSubmeshFESpace();
+  mfem::Array<int> rows( submesh_fes.GetTrueVSize() + 1 );
+  rows = 0;
+  auto mortar_tdofs_ct = 0;
+  for ( int i{ 0 }; i < submesh_fes.GetTrueVSize(); ++i ) {
+    if ( mortar_tdofs_ct < mortar_tdof_list_.Size() && mortar_tdof_list_[mortar_tdofs_ct] == i ) {
+      ++mortar_tdofs_ct;
+    }
+    rows[i + 1] = mortar_tdofs_ct;
+  }
+  // J vector
+  mfem::Array<int> mortar_tdofs( mortar_tdof_list_ );
+  // data vector
+  mfem::Vector ones( mortar_tdofs_ct );
+  ones = 1.0;
+  mfem::SparseMatrix inactive_sm( rows.GetData(), mortar_tdofs.GetData(), ones.GetData(), submesh_fes.GetTrueVSize(),
+                                  submesh_fes.GetTrueVSize(), false, false, true );
+  auto inactive_hpm = std::make_unique<mfem::HypreParMatrix>( TRIBOL_COMM_WORLD, submesh_fes.GlobalTrueVSize(),
+                                                              submesh_fes.GetTrueDofOffsets(), &inactive_sm );
+  // Have the mfem::HypreParMatrix manage the data pointers
+  rows.GetMemory().SetHostPtrOwner( false );
+  mortar_tdofs.GetMemory().SetHostPtrOwner( false );
+  ones.GetMemory().SetHostPtrOwner( false );
+  inactive_sm.SetDataOwner( false );
+  inactive_hpm->SetOwnerFlags( 3, 3, 1 );
+  block_J->SetBlock( 1, 1, inactive_hpm.release() );
+
+  return block_J;
+}
+
+std::unique_ptr<mfem::BlockOperator> MfemJacobianData::GetMfemdfdnJacobianEnzyme( const MethodData& method_data ) const
```

Comment:
Can you clarify here or elsewhere whether this routine and the one below have to be called by the user to get df/dn and dn/dx contributions? If so, is there a reason a host-code would want to or need to get the derivatives associated with the nodal normal separately? This might risk added complexity on the host code side. 

Additionally, a implicit host-code user/developer may not be as familiar with implicit contact and might just want to `getAllJacobianContributionsEnzyme()`. The way it is implemented now, that user will have to know about what dfdn or dndx means, which assumes a somewhat strong familiarity with the derivatives of the mortar method. Let's discuss.

Reply:
This will be merged in a future PR to a more general `GetMfemBlockJacobian()`, but let's keep them separate for now to test these new methods.

Users shouldn't be interacting with stuff in `MfemData.*pp`, it just exists to simplify data transfer between different meshes.  Users should be interacting with `mfem_tribol.*pp` where the user doesn't need to know the Jacobian type.

Further, since `MfemData` is just to transfer fields and Jacobians between meshes, it shouldn't know anything about Enzyme or the contact method used.